File: src/main/java/com/gmail/borlandlp/minigamesdtools/config/ConcreteConfigProvider.java
package com.gmail.borlandlp.minigamesdtools.config;

import com.gmail.borlandlp.minigamesdtools.Debug;
import com.gmail.borlandlp.minigamesdtools.config.exception.DuplicateConfigException;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

public class ConcreteConfigProvider implements ConfigProvider {
    // String[poolId] -> String[configEntityId], ConfigEntity
    private Map<String, Map<String, ConfigEntity>> configsPool = new Hashtable<>();

    public void registerEntity(ConfigPath p, ConfigEntity file) throws DuplicateConfigException {
        this.registerEntity(p.getTypeId(), file);
    }

    public void registerEntity(String poolId, ConfigEntity file) throws DuplicateConfigException {
        Map<String, ConfigEntity> configs = this.configsPool.getOrDefault(poolId, new Hashtable<>());
        if(configs.containsKey(file.getID())) {
            throw new DuplicateConfigException(poolId + " already contains " + file.getID());
        }
        configs.put(file.getID(), file);

        this.configsPool.put(poolId, configs);
    }

    public ConfigEntity getEntity(ConfigPath p, String entityId) {
        return this.getEntity(p.getTypeId(), entityId);
    }

    public ConfigEntity getEntity(String poolId, String entityId) {
        if(!this.configsPool.containsKey(poolId)) {
            Debug.print(Debug.LEVEL.WARNING, "[ConfigProvider] Detected an attempt to get entityID:" + entityId + " from unregistered poolId:" + poolId +"!");
            return null;
        } else if(!this.configsPool.get(poolId).containsKey(entityId)) {
            Debug.print(Debug.LEVEL.WARNING, "[ConfigProvider] Detected an attempt to get unknown entityID:" + entityId + " from poolId:" + poolId +"!");
            return null;
        } else {
            return this.configsPool.get(poolId).get(entityId);
        }
    }

    public List<ConfigEntity> getPoolContents(ConfigPath path) {
        return this.getPoolContents(path.getTypeId());
    }

    public List<ConfigEntity> getPoolContents(String poolId) {
        if(this.configsPool.containsKey(poolId)) {
            return new ArrayList<>(this.configsPool.get(poolId).values());
        } else {
            Debug.print(Debug.LEVEL.WARNING, "[ConfigProvider] Detected an attempt to get poolContents for unregistered poolId:" + poolId +"!");
            return new ArrayList<>();
        }
    }
}

File: src/main/kotlin/com/gmail/borlandlp/minigamesdtools/config/ConcreteConfigProvider.kt
package com.gmail.borlandlp.minigamesdtools.config

import com.gmail.borlandlp.minigamesdtools.Debug
import com.gmail.borlandlp.minigamesdtools.config.exception.DuplicateConfigException
import java.util.*

class ConcreteConfigProvider : ConfigProvider {
    // String[poolId] -> String[configEntityId], ConfigEntity
    private val configsPool: MutableMap<String, MutableMap<String, ConfigEntity?>?> =
        Hashtable()

    @Throws(DuplicateConfigException::class)
    override fun registerEntity(p: ConfigPath, file: ConfigEntity) {
        this.registerEntity(p.typeId, file)
    }

    @Throws(DuplicateConfigException::class)
    override fun registerEntity(poolId: String, file: ConfigEntity) {
        val configs =
            configsPool.getOrDefault(poolId, Hashtable<String, ConfigEntity>())!!
        if (configs.containsKey(file.id)) {
            throw DuplicateConfigException(poolId + " already contains " + file.id)
        }
        configs[file.id] = file
        configsPool[poolId] = configs
    }

    override fun getEntity(p: ConfigPath, entityId: String): ConfigEntity? {
        return this.getEntity(p.typeId, entityId)
    }

    override fun getEntity(poolId: String, entityId: String): ConfigEntity? {
        return if (!configsPool.containsKey(poolId)) {
            Debug.print(
                Debug.LEVEL.WARNING,
                "[ConfigProvider] Detected an attempt to get entityID:$entityId from unregistered poolId:$poolId!"
            )
            null
        } else if (!configsPool[poolId]!!.containsKey(entityId)) {
            Debug.print(
                Debug.LEVEL.WARNING,
                "[ConfigProvider] Detected an attempt to get unknown entityID:$entityId from poolId:$poolId!"
            )
            null
        } else {
            configsPool[poolId]!![entityId]
        }
    }

    override fun getPoolContents(path: ConfigPath): List<ConfigEntity> {
        return this.getPoolContents(path.typeId)
    }

    override fun getPoolContents(poolId: String): List<ConfigEntity> {
        return if (configsPool.containsKey(poolId)) {
            ArrayList<ConfigEntity>(configsPool[poolId]!!.values)
        } else {
            Debug.print(
                Debug.LEVEL.WARNING,
                "[ConfigProvider] Detected an attempt to get poolContents for unregistered poolId:$poolId!"
            )
            ArrayList()
        }
    }
}